Add camera module

kamtom480 opened the pull request:
This PR adds a camera module. It allows you to take a picture of different sizes.

It was tested on the Spresense board.

ladyada commented:
neat!

dhalbert commented:
@kamtom480 You can test the stub generation by doing `make stubs` in the top-level Makefile.

tannewt commented on shared-bindings/camera/Camera.c:
```diff
@@ -0,0 +1,178 @@
+
+const mp_obj_type_t camera_type = {
+   { &mp_type_type },
+   .name = MP_QSTR_GNSS,
```
```suggestion
   .name = MP_QSTR_Camera,
```

tannewt commented on py/circuitpy_defns.mk:
```diff
@@ -297,6 +300,9 @@ SRC_COMMON_HAL_ALL = \
 	busio/__init__.c \
+	camera/__init__.c \
+	camera/Camera.c \
+	camera/ImageSize.c \
```
Move this down to `SRC_BINDINGS_ENUMS` and you won't need the empty common-hal file.

tannewt commented on shared-bindings/camera/Camera.c:
```diff
@@ -0,0 +1,178 @@
+//|
+//|         cam = camera.Camera(camera.ImageSize.IMAGE_SIZE_1920x1080)
+//|
+//|         file = open("/sd/image.jpg","wb")
+//|         cam.take_picture()
+//|         file.write(cam.picture)
+//|         file.close()"""
```
I'd prefer passing in a file-like object or a bytearray. This could then be:
```suggestion
//|         with open("/sd/image.jpg","wb") as file:
//|             cam.take_picture(file)"""
```

It could also be:
```python
b = bytearray(width * height)
cam.take_picture(b)
```

What format will it produce? That seems missing in this API. I could see wanting a raw rgb565 to display on the screen for example.

Is it possible to change the image size on a per-picture basis?

dhalbert replied:
@tannewt I don't think we know the know the size of the buffer in advance if the image is compressed, so we can't necessarily pass it in. We know the max size but that may be far larger than the compressed size.

dhalbert replied:
If the low-level camera code could write directly to a stream (e.g. a file), then that would save having to allocate any RAM. But I don't know if that's possible.

kamtom480 replied:
It i possible to change the image size on a per-picture basis. For example:
```
cam = camera.Camera(camera.ImageSize.IMAGE_SIZE_2048x1536)
cam.take_picture()

cam.size = camera.ImageSize.IMAGE_SIZE_1920x1080 
cam.take_picture()

cam.size = camera.ImageSize.IMAGE_SIZE_1280x960 
cam.take_picture()
```

kamtom480 replied:
For the rest, please see my comment below.

tannewt reviewed (changes requested):
Thank you! Camera support is exciting and I agree with Dan. I'd like to see one API work for many cameras.

I'd like @PaintYourDragon to take a look as well because he's looked at the OV something something cameras too. The ESP32S2 Kaluga comes with an OV2640.

Two main questions I have are:
1) How does initialization vary with different cameras?
2) How does pixel collection vary? Is it different per-camera and per-port?

kamtom480 commented:
Thanks for all comments. I wanted to make a module that can be used for other ports. My implementation was based on this: https://learn.adafruit.com/ttl-serial-camera/circuitpython-python-usage

About `ImageSize`, I wanted the user not to have to search and type correct and supported height and width by himself, but could use the predefined sizes. I thought that in the case of more devices supporting the camera, the list of available sizes will grow. Sizes that were not supported by the camera would return an error. I thought this approach was more user-friendly. If you prefer, I can change it and the user will have to enter the height and width himself.

What do you think about the following API:
`takePicture(file)` - it takes a picture and then saves it to the file given in the argument in JPG format.
`getImage(buffer)` - returns the image from the camera to the buffer specified in the argument in RGB565 format.

Let me know what you think about it.

tannewt commented:
> Thanks for all comments. I wanted to make a module that can be used for other ports. My implementation was based on this: https://learn.adafruit.com/ttl-serial-camera/circuitpython-python-usage

Great! I hadn't seen that library so that is a great place to start.

> About `ImageSize`, I wanted the user not to have to search and type correct and supported height and width by himself, but could use the predefined sizes. I thought that in the case of more devices supporting the camera, the list of available sizes will grow. Sizes that were not supported by the camera would return an error. I thought this approach was more user-friendly. If you prefer, I can change it and the user will have to enter the height and width himself.

I think taking width and height as ints is best. We do it already for displays. Having an enum will make completion easier. However, I worry that we'd end up always having to expand the list of supported resolutions which cost code size for each variable name.

> What do you think about the following API:
> `takePicture(file)` - it takes a picture and then saves it to the file given in the argument in JPG format.
> `getImage(buffer)` - returns the image from the camera to the buffer specified in the argument in RGB565 format.

I'd rather have one function that does both. The `picamera` API for raspberry pi is another API to consider: https://picamera.readthedocs.io/en/release-1.13/recipes1.html It has one function to take a picture and also takes in a format. I don't think we should default to 'jpeg' though because not all boards may support it. I think being explicit with the format is best.

dhalbert commented:
>I think taking width and height as ints is best. We do it already for displays. Having an enum will make completion easier. However, I worry that we'd end up always having to expand the list of supported resolutions which cost code size for each variable name.

I agree. Either the resolutions supported could be documented elsewhere, or the camera API could return an optional list of resolutions. It may even be the case that some camera impls don't require fixed dimensions, so there is no list of possibilities.

There is also a difference between what the camera supports natively and what the conversion code can support. For instance, the camera might be natively 1024x768, but you could ask for a 512x384 image.

The PiCamera library @tannewt mentioned has a number of ways of specifying resolution, starting with width and height. They also let you pass in a string like "1024x768" or "QVGA", which is I think is overkill for your API.

kamtom480 commented:
I started preparing changes but I have a question. Is it possible to allocate aligned memory using CircuitPython? Something like `memalign`. The camera buffer for Spresense must be aligned.

dhalbert commented:
Heap memory is allocated in `BYTES_PER_BLOCK` chunks. Right now `BYTES_PER_BLOCK` is 16, so that should probably satisfy your alignment requirements.

jepler commented:
Does this mean that the allocations are guaranteed 16-byte aligned, though?  Supervisor allocations are in 4 byte increments, and the base address of the heap is created via supervisor allocation.

dhalbert commented:
> Does this mean that the allocations are guaranteed 16-byte aligned, though? Supervisor allocations are in 4 byte increments, and the base address of the heap is created via supervisor allocation.

I see what you mean. @kamtom480, is 4-byte alignment is good enough? If not, I think it might be good to guarantee `BYTES_PER_BLOCK` alignment for the heap, for reasons like this.

kamtom480 commented:
Spresense needs 32-byte alignment. Do you have any ideas on how to achieve this?

tannewt commented:
> Spresense needs 32-byte alignment. Do you have any ideas on how to achieve this?

Is this true for other memory allocations? Maybe we should change the GC on the Spresense port to do 32-byte blocks rather than 16.

We should also make sure the allocations for the heap are aligned to the block size too.

kamtom480 commented:
Spresense only needs a 32-byte alignment for camera. I changed `BYTES_PER_BLOCK` to 32 and aligned the heap to 32 bytes.

I also changed the API. Now `Camera` constructor has two arguments: `width` and `height`. The `take_picture` function also has two arguments: `buffer` and `format`. It also returns the size of the taken picture. The `ImageFormat` enum is created for the format.

Let me know what you think about it.

dhalbert commented on ports/cxd56/common-hal/camera/Camera.c:
```diff
@@ -0,0 +1,204 @@
+STATIC camera_dev_t camera_dev = {"/dev/video", -1};
+
+static bool camera_check_width_and_height(uint16_t width, uint16_t height) {
+    if ((width == VIDEO_HSIZE_QVGA && height == VIDEO_VSIZE_QVGA) ||
+        (width == VIDEO_HSIZE_VGA && height == VIDEO_VSIZE_VGA) ||
+        (width == VIDEO_HSIZE_HD && height == VIDEO_VSIZE_HD) ||
+        (width == VIDEO_HSIZE_QUADVGA && height == VIDEO_VSIZE_QUADVGA) ||
+        (width == VIDEO_HSIZE_FULLHD && height == VIDEO_VSIZE_FULLHD) ||
+        (width == VIDEO_HSIZE_3M && height == VIDEO_VSIZE_3M) ||
+        (width == VIDEO_HSIZE_5M && height == VIDEO_VSIZE_5M)) {
```
Not a big deal, but this could be a table of HSIZEs and VSIZEs that you iterate through to check.

dhalbert reviewed (changes requested):
Looks good! Just a few suggestions.

tannewt commented on ports/cxd56/supervisor/port.c:
```diff
@@ -100,11 +108,11 @@ uint32_t *port_stack_get_top(void) {
 
 uint32_t *port_heap_get_bottom(void) {
-    return port_stack_get_limit();
+    return heap;
```
Unfortunately I don't think this will do what you want necessarily because this heap region also contains supervisor allocations in addition to the MicroPython heap. Are you sure this is needed once the heap block size is set?

kamtom480 replied:
It is better to do it like this because now I can allocate all available memory.

tannewt commented on shared-bindings/camera/Camera.c:
```diff
@@ -0,0 +1,194 @@
+//|
+//|         cam = camera.Camera(1920, 1080)
+//|
+//|         buffer = bytearray(512 * 1024)
+//|         file = open("/sd/image.jpg","wb")
+//|         size = cam.take_picture()
+//|         file.write(buffer, size)
+//|         file.close()"""
```
Please update the example as well.

tannewt commented on shared-bindings/camera/Camera.c:
```diff
@@ -0,0 +1,194 @@
+//|     def take_picture(self, buf: WriteableBuffer, format: ImageFormat) -> int:
+//|         """Take picture and save to ``buf`` in the given ``format``
+//|
+//|         :return: the size of the picture taken
```
```suggestion
//|         :return: the number of bytes written into buf
```

tannewt commented on shared-bindings/camera/__init__.c:
```diff
@@ -0,0 +1,50 @@
+    { MP_ROM_QSTR(MP_QSTR_Camera), MP_ROM_PTR(&camera_type) },
+
+    // Enum-like Classes.
+    { MP_ROM_QSTR(MP_QSTR_ImageSize), MP_ROM_PTR(&camera_imageformat_type) },
```
```suggestion
    { MP_ROM_QSTR(MP_QSTR_ImageFormat), MP_ROM_PTR(&camera_imageformat_type) },
```

tannewt reviewed (changes requested):
Is there a significant delay between starting the camera and being able to take a picture? I'm wondering why we can't just pass size into `take_picture` too.

kamtom480 commented on ports/cxd56/supervisor/port.c:
```diff
@@ -46,12 +46,20 @@
 #include "common-hal/pwmio/PWMOut.h"
 #include "common-hal/busio/UART.h"
 
+#define HEAP_SIZE (1000 * 1024)
```
I changed it. Now I allocate all available memory.

tannewt reviewed (commented):
This looks good to me! I'd wait a day to give @Andon-A a chance to comment on the API. He is working on https://github.com/adafruit/Adafruit_CircuitPython_VC0706/pull/18

tannewt commented:
@kamtom480 Feel free to mark comments as resolved as you do them instead of reacting with :+1:

Andon-A commented:
OK, so, I've had a short look. I'm going to preface this with *I am **completely** out of my league* as far as this type of programming goes, and am primarily going off of how the usage example looks. I expect I'm going to make some assumptions that are wrong, so please do correct me!

I'm going to assume there's something to differentiate between how the camera is interfaced (SPI, I2C, UART, whatever) during initialization.

The VC0706 in particular, though, can take pictures at 640x480, 320x240, and 160x120, and can flip between those on a whim. It *looks* like your initializing it at one set size. This... works? but can be restrictive if, say, you want to take a smaller picture to transmit it over a packet radio faster. but also take larger ones to save to local storage. Again, a lot of this is stuff I quite frankly don't know enough about to know what I'm looking at, but I didn't see anything that implied it was able to change size if the camera allowed it.

Another issue with the VC0706 is how you have to save it. You have `cam.take_picture` followed by simply writing that. With the VC0706, you actually pause the buffer and then read the buffer a few (Generally 32) bytes at a time, saving as you go. A single bulk save wouldn't exactly work with this camera. Additionally, the VC0706 doesn't automatically resume being a video source until you tell it to be a video source. Which can't be done until after you've finished saving the image. I honestly don't know what other types of cameras there are or how they behave, so this might be normal and something that'll be done, or it might be weird out there bizzar and has to have a special case.

The VC0706 also has some controls for zoom (Which looks like literally scaling up the image like you would in Photoshop) and pan (Which moves around said scaled up image if you're zoomed), along with some gain, brightness, and some other controls which I'm not even sure are implemented in the Arduino library. Some method to interact with such extra controls for cameras that have them wouldn't be amiss - but again, I haven't done a deep dive on the code so it might already be there.

kamtom480 commented:
@Andon-A Thank you for your comment!

> The VC0706 in particular, though, can take pictures at 640x480, 320x240, and 160x120, and can flip between those on a whim. It _looks_ like your initializing it at one set size. This... works? but can be restrictive if, say, you want to take a smaller picture to transmit it over a packet radio faster. but also take larger ones to save to local storage. Again, a lot of this is stuff I quite frankly don't know enough about to know what I'm looking at, but I didn't see anything that implied it was able to change size if the camera allowed it.

A `Camera` class has two variables: `width` and `height`. Thanks to them you can change the size of the picture. Example:
```
cam = camera.Camera(1920, 1080)
cam.take_picture(buffer, camera.ImageFormat.JPG) 
cam.width = 320
cam.height = 240
cam.take_picture(buffer, camera.ImageFormat.JPG)
```

> Another issue with the VC0706 is how you have to save it. You have `cam.take_picture` followed by simply writing that. With the VC0706, you actually pause the buffer and then read the buffer a few (Generally 32) bytes at a time, saving as you go. A single bulk save wouldn't exactly work with this camera. Additionally, the VC0706 doesn't automatically resume being a video source until you tell it to be a video source. Which can't be done until after you've finished saving the image. I honestly don't know what other types of cameras there are or how they behave, so this might be normal and something that'll be done, or it might be weird out there bizzar and has to have a special case.

I think you can implement reading and writing the image to buffer in the `take_picture()` function. You can return from this function after filling the buffer with an image.

> The VC0706 also has some controls for zoom (Which looks like literally scaling up the image like you would in Photoshop) and pan (Which moves around said scaled up image if you're zoomed), along with some gain, brightness, and some other controls which I'm not even sure are implemented in the Arduino library. Some method to interact with such extra controls for cameras that have them wouldn't be amiss - but again, I haven't done a deep dive on the code so it might already be there.

Spresense also has other camera features. However, at the beginning I wanted to add a basic camera support and then expand it later.

tannewt commented:
@kamtom480 Could we pass a size into `take_picture` instead of passing it earlier? Would that cause to long of a delay?

kamtom480 commented:
> Could we pass a size into `take_picture` instead of passing it earlier? Would that cause to long of a delay?

@tannewt We can pass the size in `take_picture`. There is no problem with that. The API will look like this:
```
cam = camera.Camera()
cam.take_picture(1920, 1080, buffer, camera.ImageFormat.JPG)
```
Should I change it?

Andon-A commented:
Does the camera class *need* two separate variables for width and height? I know many (standard) cameras have locked sizes, so having both be passed separately seems strange on the surface to me

tannewt commented:
I'd do:
```python
cam.take_picture(buffer, width=1920, height=1080, format=camera.ImageFormat.JPG)
```

@Andon-A I like having them separate because you don't need a tuple then to store them and you can also "label" them by making them keyword only arguments.

Andon-A commented:
Makes sense! And you don't have to pile consts and stuff to get there either.

tannewt reviewed (approved):
This looks great to me! Thank you!

dhalbert commented:
Looks good to me too!
